soc: arm: cypress: psoc6: Add Cortex-M0+ interrupt mux support

nandojve opened the pull request:
PSoC-6 SoC needs that user define the nvic interrupt number to bind with the peripheral interrupt line for the Cortex-M0+ CPU.  It uses a multiplex before any NVIC interrupt line.  The minimal support was added to allow interrupts be available for both CPUs.

Note: The PSoC-6 SoC allows that both CPUs receive the same interrupt. A tipical use is GPIO interrupt handle and user is responsable to define interrupt line, priority and take care of enable same peripheral instance on both CPUs only when appropriated.

This update <soc.h> include files removing the unnecessary <kernel_includes.h> file. In addition, add <sys/util.h> to expose macros and <devicetree.h> following general standards.

This solves #28442 and allows development of drivers that can be shared between SoC cores including, external interrupts by GPIO, serial, spi etc.

nandojve committed:
soc: arm: cypress: psoc6: Add Cortex-M0+ int mux support

PSoC-6 SoC needs that user define the nvic interrupt number to bind
with the peripheral interrupt line for the Cortex-M0+ CPU.  It uses
a multiplex before any NVIC interrupt line.  The interrupt vector is
selected using interrupt-parent property with the intmux_chN number
reference.

Note: The PSoC-6 SoC allows that both CPUs receive the same interrupt.
A tipical use is GPIO interrupt handle and user is responsable to
define interrupt line, priority and take care of enable same peripheral
instance on both CPUs only when appropriated.

Signed-off-by: Gerson Fernando Budke <gerson.budke@atl-electronics.com>

nandojve committed:
soc: arm: cypress: psoc6: Update soc.h includes

Update <soc.h> include files.  This removes the unnecessary
<kernel_includes.h> file.  In addition, add <sys/util.h> to
expose macros and <devicetree.h> following general standards.

Signed-off-by: Gerson Fernando Budke <gerson.budke@atl-electronics.com>

nandojve commented:
Rebase and add missing **map-nvic** documentation.

galak commented:
Need to look at this in more detail.   Wondering how this compares to what we have for rv32m1_vega

nandojve commented:
> Need to look at this in more detail. Wondering how this compares to what we have for rv32m1_vega

I know this is an intermediate solution. At drivers, the work was done on the macro that connects IRQ and it was defined externally. @mbolivar already told me about RISC-V but I didn't see how user define the interrupt.

mbolivar-nordic commented:
> I didn't see how user define the interrupt on DT, Cy m0 problem.

I don't understand "Cy m0 problem", sorry.

Perhaps this is a chance to clean up the devicetree documentation (and maybe implementation) with regards to multi-level interrupts, but can you explain what you tried when you tried to make intermediate interrupt controller nodes for the interrupt mux? The idea is to set the `interrupt-parent` to point to those and to use the normal `interrupts` property to specify interrupts in DT. Can you explain what didn't work the way you expected when you tried that?

nandojve committed:
dts: arm: cypress: psoc6: Drop nvic reference

The psoc6.dtsi file declare a reference to nvic.  Since it was proper
defined at psoc6_cm0/4.dtsi files this entry is redundant.  Drop the
useless entry.

Signed-off-by: Gerson Fernando Budke <gerson.budke@atl-electronics.com>

nandojve commented:
> > I didn't see how user define the interrupt on DT, Cy m0 problem.
> 
> I don't understand "Cy m0 problem", sorry.
> 
> Perhaps this is a chance to clean up the devicetree documentation (and maybe implementation) with regards to multi-level interrupts, but can you explain what you tried when you tried to make intermediate interrupt controller nodes for the interrupt mux? The idea is to set the `interrupt-parent` to point to those and to use the normal `interrupts` property to specify interrupts in DT. Can you explain what didn't work the way you expected when you tried that?

I think I understand my difficult, for some reason I didn't look at some file on rv32m1_vega dts directory. That create a gap that I wasn't understand how to fill. I hope this update is close to what I rely need to do.
Thanks!

mbolivar-nordic commented on dts/arm/cypress/psoc6_cm0.dtsi:
```diff
@@ -15,6 +15,306 @@
+				reg = <0x0>;
+				#interrupt-cells = <2>;
+				interrupt-controller;
+				interrupts = <0 0>;
```
It's your decision, but do you think you might want to make the default priority a bit lower (assuming 0 means "highest priority") for these?

We did that recently for Nordic SoCs to make it easier for users to increase the priority of just some interrupts without having to change priorities for other unrelated nodes.

nandojve replied:
fully agreed, thanks!

mbolivar-nordic reviewed (approved):
This looks nice to me! Just a couple of minor questions but I think this seems sane.

nandojve commented on soc/arm/cypress/common/cypress_psoc6_dt.h:
```diff
@@ -0,0 +1,46 @@
+#ifdef CONFIG_CPU_CORTEX_M0PLUS
+#define DO_NOTHING()
+#define CY_PSOC6_DT_NVIC_MUX_INSTALL(n, isr, label)	    \
+	COND_CODE_1(DT_INST_NODE_HAS_PROP(n, interrupt_parent),\
```
Yes, fixed!

nandojve commented:
rebase

nandojve commented:
Hello @galak !
I was wondering if there is anything missing here?

ADEscobar commented:
Is this ready to be merged then?

mbolivar-nordic commented:
> Is this ready to be merged then?

No, it needs another +1, ideally from @galak

galak commented on dts/arm/cypress/psoc6_cm0.dtsi:
```diff
@@ -15,6 +15,306 @@
+			label = "INTMUX";
+			status = "disabled";
+			#address-cells = <1>;
+			#size-cells = <0>;
```
Let's make `#size-cells = <1>;`

nandojve replied:
fixed!

galak reviewed (changes requested):
* Change #size-cells to 1 for mux registers.  Even if size is just 1, since these are still MMIO registers.
* Add some comments on what / how the mux values are figured out.  Seen inline.

galak commented on dts/bindings/interrupt-controller/cypress,psoc6-intmux-ch.yaml:
```diff
@@ -0,0 +1,23 @@
+# SPDX-License-Identifier: Apache-2.0
+
+description: Cypress Interrupt Multiplex Channel
+
```
Can we add more description either here or in the `cypress,psoc6-intmux` binding on how this works and is used by a user.  ie the bits of setting the `interrupt-parent` for a device that is targeted to the M0+ to once of the `cypress,psoc6-intmux-ch`, etc.

nandojve replied:
Add documentation at cypress,psoc6-int-mux.yaml

galak commented on soc/arm/cypress/common/cypress_psoc6_dt.h:
```diff
@@ -0,0 +1,50 @@
+
+/* Devicetree macros related to interrupt */
+#ifdef CONFIG_CPU_CORTEX_M0PLUS
+#define CY_PSOC6_DT_NVIC_MUX_INSTALL(n, isr, label)	    \
```
Can you add a comment block here on what `n`, `isr`, and `label` are.

galak replied:
Would also rename and maybe drop the 'NVIC_MUX' part, since as I understand it any cy6 driver should use the macros regardless.

galak commented on soc/arm/cypress/common/cypress_psoc6_dt.h:
```diff
@@ -0,0 +1,50 @@
+#define CY_PSOC6_DT_NVIC_MUX_MAP(n)
+#endif
+
+#define CY_PSOC6_NVIC_MUX_INSTALL(n, isr, label)	  \
```
* Rename to `CY_PSOC6_IRQ_CONFIG`
* Added comment block for what `n`, `isr`, `label` are.

nandojve replied:
Fixed with docs.

galak commented on soc/arm/cypress/common/cypress_psoc6_dt.h:
```diff
@@ -0,0 +1,50 @@
+#define _CYPRESS_PSOC6_DT_H_
+
+#include <devicetree.h>
+
```
Add a comment here about the main "API" being `CY_PSOC6_IRQ_CONFIG`.

nandojve replied:
Addressed in the docs.

galak commented on soc/arm/cypress/common/cypress_psoc6_dt.h:
```diff
@@ -0,0 +1,101 @@
+ * - configure interrupt channel using the channel number register value from
+ *   interrupt-parent node.
+ */
+#define CY_PSOC6_DT_NVIC_INSTALL(n, isr)	              \
```
Based on what we are doing here, we should do `_DT_INST_`.

nandojve replied:
Fixed!

galak commented on dts/bindings/interrupt-controller/cypress,psoc6-int-mux.yaml:
```diff
@@ -0,0 +1,72 @@
+  user can select up to 32 interrupts sources from the 240 possible vectors
+  to be processed in the Cortex-M0+ CPU.
+
+  At CPU Sub System (CPUSS) there are 8 special registers to configure the
```
Can you add a sentence or something here that the 'vector number' is that get placed into the mux field is the IRQ number for the NVIC to the Cortex-M4.

nandojve replied:
Done!

galak commented on dts/arm/cypress/psoc6_cm0.dtsi:
```diff
@@ -15,6 +15,307 @@
+			compatible = "cypress,psoc6-intmux";
+			reg = <0x40210020 0x20>;
+			label = "INTMUX";
+			status = "disabled";
```
should probably remove `status="disabled"` here.

nandojve replied:
yes, it will be easy for user - switch to "okay"

galak commented on dts/arm/cypress/psoc6_cm0.dtsi:
```diff
@@ -15,6 +15,307 @@
+			label = "INTMUX";
+			status = "disabled";
+			#address-cells = <1>;
+			#size-cells = <1>;
```
technically we should have:

`ranges = <0x0 0x40210020 0x20>;`

(this will change the reg value to be the correct MMIO address, I don't think it impacts anything as it doesn't look like you are using the reg addr).

nandojve replied:
Done!

galak reviewed (changes requested):
Some minor tweaks, and on general question if we should just remove `status="disabled"` for the "cypress,psoc6-intmux-ch" nodes.  Means one less thing the board dts file would have to modify.

galak commented:
One other nit: can you see if we can remove `label`.  Want to phase it out in the future and I don't think you need it at all.

galak commented on soc/arm/cypress/common/cypress_psoc6_dt.h:
```diff
@@ -0,0 +1,101 @@
+#define CY_PSOC6_NVIC_MUX_MAP(n)
+#endif
+
+#define CY_PSOC6_IRQ_CONFIG(n, isr)			\
```
CY_PSOC6_IRQ_CONFIG -> CY_PSOC6_DT_INST_IRQ_CONFIG

nandojve replied:
I fixed to have only the public interface with _DT_INST_ in the middle following the <manufacturer prefix>_DT[_INST]_macro_name. After I read this I remember that already made this in the past.

nandojve commented:
> One other nit: can you see if we can remove `label`. Want to phase it out in the future and I don't think you need it at all.

Yes, I removed and tested with GPIO driver which depends on this.
